From 8b60039133e7e3158c6425853f6a3e8d5bca35b1 Mon Sep 17 00:00:00 2001
From: Dmitry Shifrin <dmitry.shifrin@cogentembedded.com>
Date: Thu, 29 Sep 2016 12:07:52 +0300
Subject: [PATCH] Fix adv7511 drm driver Fixes which makes driver work with
 GreenHouse monitor: 1. Allow interrupts 2. Use power functions instead of
 copy it's code 3. Fix encoder_detect function

---
 drivers/gpu/drm/i2c/adv7511.c | 112 ++++++++++++++++++++++++++++++++----------
 1 file changed, 85 insertions(+), 27 deletions(-)

diff --git a/drivers/gpu/drm/i2c/adv7511.c b/drivers/gpu/drm/i2c/adv7511.c
index 00416f2..76fc4ba 100644
--- a/drivers/gpu/drm/i2c/adv7511.c
+++ b/drivers/gpu/drm/i2c/adv7511.c
@@ -20,6 +20,8 @@
 
 #include "adv7511.h"
 
+#define MAX_EDID_ITERATIONS		2
+
 struct adv7511 {
 	struct i2c_client *i2c_main;
 	struct i2c_client *i2c_edid;
@@ -33,7 +35,7 @@ struct adv7511 {
 
 	unsigned int current_edid_segment;
 	uint8_t edid_buf[256];
-	bool edid_read;
+	bool edid_read, read_error;
 
 	wait_queue_head_t wq;
 	struct drm_encoder *encoder;
@@ -386,6 +388,11 @@ static void adv7511_power_on(struct adv7511 *adv7511)
 	 */
 	regcache_sync(adv7511->regmap);
 
+	regmap_write(adv7511->regmap, ADV7511_REG_INT_ENABLE(0),
+				ADV7511_INT0_EDID_READY | ADV7511_INT0_HDP);
+	regmap_write(adv7511->regmap, ADV7511_REG_INT_ENABLE(1),
+				ADV7511_INT1_DDC_ERROR);
+
 	adv7511->powered = true;
 }
 
@@ -400,6 +407,13 @@ static void adv7511_power_off(struct adv7511 *adv7511)
 	adv7511->powered = false;
 }
 
+static void adv7511_reset(struct adv7511 *adv7511)
+{
+	adv7511_power_off(adv7511);
+	msleep(5);
+	adv7511_power_on(adv7511);
+}
+
 /* -----------------------------------------------------------------------------
  * Interrupt and hotplug detection
  */
@@ -442,7 +456,8 @@ static int adv7511_irq_process(struct adv7511 *adv7511)
 		drm_helper_hpd_irq_event(adv7511->encoder->dev);
 
 	if (irq0 & ADV7511_INT0_EDID_READY || irq1 & ADV7511_INT1_DDC_ERROR) {
-		adv7511->edid_read = true;
+		adv7511->edid_read = (irq0 & ADV7511_INT0_EDID_READY) != 0;
+		adv7511->read_error = (irq1 & ADV7511_INT1_DDC_ERROR) != 0;
 
 		if (adv7511->i2c_main->irq)
 			wake_up_all(&adv7511->wq);
@@ -463,27 +478,56 @@ static irqreturn_t adv7511_irq_handler(int irq, void *devid)
 /* -----------------------------------------------------------------------------
  * EDID retrieval
  */
+static void adv7511_start_read_edid_block(struct adv7511 *adv7511, unsigned int block)
+{
+	adv7511->edid_read = false;
+	adv7511->read_error = false;
+	regmap_write(adv7511->regmap, ADV7511_REG_EDID_SEGMENT,
+				     block);
+}
 
-static int adv7511_wait_for_edid(struct adv7511 *adv7511, int timeout)
+static int adv7511_wait_for_edid(struct adv7511 *adv7511, int timeout, 
+					unsigned int block, bool restart)
 {
 	int ret;
+	int i;
+
+	for (i = 0; i < MAX_EDID_ITERATIONS; i++)
+	{
+		if (restart)
+			adv7511_start_read_edid_block(adv7511, block);
+
+		restart = true;
+
+		if (adv7511->i2c_main->irq) {
+			ret = wait_event_interruptible_timeout(adv7511->wq,
+					(adv7511->edid_read || adv7511->read_error), msecs_to_jiffies(timeout));
+
+			if (ret == 0)
+				return -EAGAIN;
 
-	if (adv7511->i2c_main->irq) {
-		ret = wait_event_interruptible_timeout(adv7511->wq,
-				adv7511->edid_read, msecs_to_jiffies(timeout));
-	} else {
-		for (; timeout > 0; timeout -= 25) {
-			ret = adv7511_irq_process(adv7511);
 			if (ret < 0)
-				break;
+				return -EIO;
+		} else {
+			for (; timeout > 0; timeout -= 25) {
+				ret = adv7511_irq_process(adv7511);
+				if (ret < 0)
+					break;
 
-			if (adv7511->edid_read)
-				break;
+				if (adv7511->edid_read || adv7511->read_error)
+					break;
 
-			msleep(25);
+				msleep(25);
+			}
 		}
-	}
 
+		if (adv7511->read_error)
+			adv7511_reset(adv7511);
+
+
+		if (adv7511->edid_read)
+			break;
+	}
 	return adv7511->edid_read ? 0 : -EIO;
 }
 
@@ -500,18 +544,24 @@ static int adv7511_get_edid_block(void *data, u8 *buf, unsigned int block,
 		return -EINVAL;
 
 	if (adv7511->current_edid_segment != block / 2) {
-		unsigned int status;
+		unsigned int status, blk;
 
 		ret = regmap_read(adv7511->regmap, ADV7511_REG_DDC_STATUS,
 				  &status);
 		if (ret < 0)
 			return ret;
 
-		if (status != 2) {
-			adv7511->edid_read = false;
-			regmap_write(adv7511->regmap, ADV7511_REG_EDID_SEGMENT,
-				     block);
-			ret = adv7511_wait_for_edid(adv7511, 200);
+		if (!(status & 0x0f))
+			return -EIO;
+
+		/* timeout situation */
+		ret = regmap_read(adv7511->regmap, ADV7511_REG_EDID_SEGMENT,
+				  &blk);
+		if (ret < 0)
+			return ret;
+
+		if (status != 2 || blk != block / 2) {
+			ret = adv7511_wait_for_edid(adv7511, 200, block, blk != block / 2);
 			if (ret < 0)
 				return ret;
 		}
@@ -564,24 +614,29 @@ static int adv7511_get_modes(struct drm_encoder *encoder,
 	struct adv7511 *adv7511 = encoder_to_adv7511(encoder);
 	struct edid *edid;
 	unsigned int count;
+	bool powered = adv7511->powered;
 
 	/* Reading the EDID only works if the device is powered */
-	if (!adv7511->powered) {
-		regmap_write(adv7511->regmap, ADV7511_REG_INT(0),
+	if (!powered) {
+		/*regmap_write(adv7511->regmap, ADV7511_REG_INT(0),
 			     ADV7511_INT0_EDID_READY);
 		regmap_write(adv7511->regmap, ADV7511_REG_INT(1),
 			     ADV7511_INT1_DDC_ERROR);
 		regmap_update_bits(adv7511->regmap, ADV7511_REG_POWER,
 				   ADV7511_POWER_POWER_DOWN, 0);
-		adv7511->current_edid_segment = -1;
-	}
+		adv7511->current_edid_segment = -1;*/
 
+		/* This call seems more correct */
+		adv7511_power_on(adv7511);
+	}
 	edid = drm_do_get_edid(connector, adv7511_get_edid_block, adv7511);
 
-	if (!adv7511->powered)
-		regmap_update_bits(adv7511->regmap, ADV7511_REG_POWER,
+	if (!powered) {
+		/*regmap_update_bits(adv7511->regmap, ADV7511_REG_POWER,
 				   ADV7511_POWER_POWER_DOWN,
-				   ADV7511_POWER_POWER_DOWN);
+				   ADV7511_POWER_POWER_DOWN);*/
+		adv7511_power_off(adv7511);
+	}
 
 	kfree(adv7511->edid);
 	adv7511->edid = edid;
@@ -635,8 +690,11 @@ adv7511_encoder_detect(struct drm_encoder *encoder,
 		regcache_mark_dirty(adv7511->regmap);
 		adv7511_power_on(adv7511);
 		adv7511_get_modes(encoder, connector);
+		/* it seems starnge returning disconnect whet it is connected */
+		#if 0
 		if (adv7511->status == connector_status_connected)
 			status = connector_status_disconnected;
+		#endif
 	} else {
 		/* Renable HDP sensing */
 		regmap_update_bits(adv7511->regmap, ADV7511_REG_POWER2,
-- 
1.9.1

